File: test/src/test/java/com/ciyfhx/test/ClientTest.java
/*
 * Copyright (c) 2018.
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.ciyfhx.test;

import java.util.Arrays;


import com.ciyfhx.network.*;
import com.ciyfhx.network.validator.MACValidator;
import com.ciyfhx.processors.Processors;
import com.ciyfhx.processors.TransformProcessor;
import com.ciyfhx.test.packet.PacketIDs;
import java8.util.concurrent.SubmissionPublisher;
import org.slf4j.impl.SimpleLogger;

public class ClientTest {

	public static void main(String[] args) throws Exception, PacketsFactory.NotFoundException {
		
		System.setProperty(SimpleLogger.DEFAULT_LOG_LEVEL_KEY, "TRACE");

		PacketsFactory factory = new PacketsFactory();

		factory.registerIds(Arrays.asList(PacketIDs.MESSAGING));
		SubmissionPublisher<PacketEvent<Packet>> publisher = factory.getPublisher(PacketIDs.MESSAGING);

		TransformProcessor<PacketEvent, String> stringTransformProcessor = Processors.ToStringProcessor;

		publisher.subscribe(stringTransformProcessor);
		stringTransformProcessor.subscribe(new PrintLineSubscriber());

		Client client = ClientBuilder.newInstance().withPacketsFactory(factory).build();

		client.setNetworkListener(new NetworkListener() {
			@Override
			public void disconnected(NetworkConnection disconnector) {
				System.out.println("Disconnector: " + disconnector.getAddress());
			}

			@Override
			public void connected(NetworkConnection connector) {
				System.out.println("Connector: " + connector.getAddress());
			}
		});

		client.connectAsync("localhost", 5555).thenAccept((con) -> {
			con.getPipeLineStream().addPipeLine(new CompressionPipeLine());
		});

		//Blocking
		while(true){

		}

		//client.sendPacket(new MessagingPacket("Hello"));

	}

}

File: test/src/test/kotlin/com/ciyfhx/test/ClientTest.kt


package com.ciyfhx.test

import com.ciyfhx.network.ClientBuilder
import com.ciyfhx.network.Packet
import com.ciyfhx.network.PacketsFactory
import com.ciyfhx.network.authenticate.AuthenticationManager
import com.ciyfhx.network.authentication.AuthenticationManagerList
import com.ciyfhx.network.authentication.BasicAuthenticationManager
import com.ciyfhx.network.authentication.DigestAuthenticationManager
import com.ciyfhx.network.authentication.credential
import com.ciyfhx.network.build
import com.ciyfhx.processors.Processors
import java8.util.concurrent.Flow
import org.slf4j.impl.SimpleLogger
import java.nio.ByteBuffer
import javax.net.ssl.SSLContext


const val MESSAGE = 1

data class MessagePacket(val message: String) : Packet(MESSAGE, ByteBuffer.wrap(message.toByteArray()))

class PrintLineSubscriber : Flow.Subscriber<String> {
    lateinit var subscription: Flow.Subscription
    override fun onSubscribe(subscription: Flow.Subscription) {
        this.subscription = subscription
        println("Subscribe to $subscription")
        subscription.request(1)
    }

    override fun onNext(item: String) {
        println("Message: $item")         
        subscription.request(1)
    }

    override fun onError(throwable: Throwable) = subscription.cancel()

    override fun onComplete() =  println("Done")

}

fun main(args: Array<String>){

    System.setProperty("javax.net.ssl.trustStore", "D:\\keystore.jks")
    System.setProperty(SimpleLogger.DEFAULT_LOG_LEVEL_KEY, "TRACE")
    val packageFactory = PacketsFactory()
    val publisher = packageFactory.registerId(MESSAGE)
    val toStringProcessor = Processors.ToStringProcessor
    publisher.subscribe(toStringProcessor)
    toStringProcessor.subscribe(PrintLineSubscriber())


    val authenticationManager = AuthenticationManagerList(AuthenticationManager.getDefaultAuthenticationManager(),
            DigestAuthenticationManager("hello" credential "123".toCharArray(), "com.1ciyfhx.test"))
    val client = ClientBuilder.newInstance().build(packetsFactory = packageFactory)//authenticationManager = authenticationManager
    client.connectAsync("192.168.1.229", 5556).thenAccept {
        println("Client Connected")
    }.exceptionally {
        it.printStackTrace()
        null
    }

    while(true){
        Thread.sleep(1000)
    }

}

